### PR TITLE
changes to support SCTP LB

### PR DESCRIPTION
**Pull request by thakurajayL**
Changes to keep SCTP connection management in separate POD.

### CODE REVIEW

File: service/init.go [hunk 1]
```diff
@@ -335,6 +336,7 @@ func (amf *AMF) Start() {
 	ngap_service.Run(self.NgapIpList, 38412, ngapHandler)
 
 	go amf.SendNFProfileUpdateToNrf()
+    go grpcServer.Start_service()
```
Comment (thakurajayL) [inline service/init.go:339]: formatting required.

File: ngap/dispatcher.go [hunk 2]
```diff
@@ -39,6 +49,7 @@ func Dispatch(conn net.Conn, msg []byte) {
 	pdu, err := ngap.Decoder(msg)
 	if err != nil {
 		ran.Log.Errorf("NGAP decode error : %+v", err)
+        fmt.Println("DispatchLb, decode Messgae error ", remoteAddr)
```
Comment (thakurajayL) [inline ngap/dispatcher.go:52]: Remove println and use logs

File: ngap/dispatcher.go [hunk 3]
```diff
@@ -18,19 +18,29 @@ import (
...
+    fmt.Println("Conn %v and msg %v ", conn, msg)
+}
+
+func DispatchLb(remoteAddr string, msg []byte, Amf2RanMsgChan chan *sdcoreAmfServer.Message) {
+    fmt.Printf("DispatchLb %v %T", remoteAddr, Amf2RanMsgChan)
 	var ran *context.AmfRan
 	amfSelf := context.AMF_Self()
 
-	ran, ok := amfSelf.AmfRanFindByConn(conn)
+	ran, ok := amfSelf.AmfRanFindByConn(remoteAddr) /* ajay */
```
Comment (thakurajayL) [inline ngap/dispatcher.go:34]: Remote address should be gNB id or gNB ip address.

File: ngap/dispatcher.go [hunk 4]
```diff
@@ -193,13 +204,17 @@ func DispatchNgapMsg(ran *context.AmfRan, pdu *ngapType.NGAPPDU) {
...
 func HandleSCTPNotification(conn net.Conn, notification sctp.Notification) {
-	amfSelf := context.AMF_Self()
+    fmt.Println("Conn %v notification %v ", conn, notification)
+}
+func HandleSCTPNotificationSctp(remoteAddr string, notification sctp.Notification) {
 
-	logger.NgapLog.Infof("Handle SCTP Notification[addr: %+v]", conn.RemoteAddr())
+	logger.NgapLog.Infof("Handle SCTP Notification[addr: %+v]", remoteAddr)
 
+/*
```
Comment (thakurajayL) [inline ngap/dispatcher.go:213]: Need to uncomment this code ?
 - How handle Notification Sctp is called ?

Comment (thakurajayL) [reply ngap/dispatcher.go:213]: SCTP POD should inform if gNB is going down and then AMF POD can invoke this notification function.

File: ngap/dispatcher.go [hunk 5]
```diff
@@ -193,13 +204,17 @@ func DispatchNgapMsg(ran *context.AmfRan, pdu *ngapType.NGAPPDU) {
 }
 
 func HandleSCTPNotification(conn net.Conn, notification sctp.Notification) {
-	amfSelf := context.AMF_Self()
+    fmt.Println("Conn %v notification %v ", conn, notification)
+}
+func HandleSCTPNotificationSctp(remoteAddr string, notification sctp.Notification) {
```
Comment (thakurajayL) [inline ngap/dispatcher.go:209]: define message type and sctp-pod should send this event to AMF.

File: gmm/handler.go [hunk 6]
```diff
@@ -1294,7 +1294,7 @@ func handleRequestedNssai(ue *context.AmfUe, anType models.AccessType) error {
 					UserLocation:     &ue.Location,
 					RrcEstCause:      ue.RanUe[anType].RRCEstablishmentCause,
 					UeContextRequest: ue.RanUe[anType].UeContextRequest,
-					AnN2IPv4Addr:     ue.RanUe[anType].Ran.Conn.RemoteAddr().String(),
+					AnN2IPv4Addr:     ue.RanUe[anType].Ran.RemoteAddr,
```
Comment (thakurajayL) [inline gmm/handler.go:1297]: Need to make sure Ran Remote address is correct.

Comment (thakurajayL) [reply gmm/handler.go:1297]: we should get remote address and ran id separately. Looks like we need both fields.

Comment (thakurajayL) [reply gmm/handler.go:1297]: Change remoteaddr variable name

File: context/context.go [hunk 7]
```diff
@@ -292,19 +291,18 @@ func (context *AMFContext) AmfUeDeleteBySuci(suci string) (ue *AmfUe, ok bool) {
 	return
 }
 
-func (context *AMFContext) NewAmfRan(conn net.Conn) *AmfRan {
+func (context *AMFContext) NewAmfRan(remoteAddr string) *AmfRan {
```
Comment (thakurajayL) [inline context/context.go:294]: Use gnbId instead of address.

File: context/context.go [hunk 8]
```diff
@@ -292,19 +291,18 @@ func (context *AMFContext) AmfUeDeleteBySuci(suci string) (ue *AmfUe, ok bool) {
...
+	ran.GnbIp = remoteAddr
+	ran.Log = logger.NgapLog.WithField(logger.FieldRanAddr, remoteAddr)
+	context.AmfRanPool.Store(remoteAddr, &ran)
 	return &ran
 }
 
 // use net.Conn to find RAN context, return *AmfRan and ok bit
-func (context *AMFContext) AmfRanFindByConn(conn net.Conn) (*AmfRan, bool) {
-	if value, ok := context.AmfRanPool.Load(conn); ok {
+func (context *AMFContext) AmfRanFindByConn(remoteAddr string) (*AmfRan, bool) {
```
Comment (thakurajayL) [inline context/context.go:304]: Keep 2 separate functions. findbyconn and findbyId

File: context/context.go (hunk 7 above)
Comment (thakurajayL) [reply context/context.go:294]: Keep original support of creating NewAmfRan. This helps in running AMF without sctplb

File: context/amf_ran.go [hunk 9]
```diff
@@ -92,6 +95,7 @@ func (ran *AmfRan) RemoveAllUeInRan() {
 }
 
 func (ran *AmfRan) RanUeFindByRanUeNgapID(ranUeNgapID int64) *RanUe {
+    // ajay - need fix.. we xan not have serial search
```
Comment (thakurajayL) [inline context/amf_ran.go:98]: remove name and add TODO comment.

File: amfGrpcServer/server.proto [hunk 10]
```diff
@@ -0,0 +1,25 @@
+syntax = "proto3";
+package sdcoreAmfServer;
+option go_package = "./sdcoreAmfServer";
+
+enum msgType {
+    UNKNOWN   = 0;
```
Comment (thakurajayL) [inline amfGrpcServer/server.proto:6]: format please

File: amfGrpcServer/server.proto [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+	INIT_MSG  = 1;
+	GNB_MSG   = 2;
+    AMF_MSG   = 3;
+	GNB_DISC  = 4;
+	GNB_CONN  = 5;
+}
+
+message Message {
+  string LbId = 1; 
+  string ServerId = 2;
```
Comment (thakurajayL) [inline amfGrpcServer/server.proto:16]: Let't change serverId to amfid.

File: amfGrpcServer/server.proto [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+	GNB_DISC  = 4;
+	GNB_CONN  = 5;
+}
+
+message Message {
+  string LbId = 1; 
+  string ServerId = 2; 
+  msgType Msgtype = 3;
+  string GnbId = 4; 
+  string body = 5;
```
Comment (thakurajayL) [inline amfGrpcServer/server.proto:19]: Review this name and check how it works when looking at sctplb code.

File: amfGrpcServer/server.proto [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  bytes msg = 6;
+}
+
+service ChatService {
```
Comment (thakurajayL) [inline amfGrpcServer/server.proto:23]: Change name.

File: amfGrpcServer/amf_server.go [hunk 14]
```diff
@@ -0,0 +1,82 @@
+package amf_server
```
Comment (thakurajayL) [inline amfGrpcServer/amf_server.go:1]: this file should be moved to main package.
and change the directory name amfGrpcServer--> protos

File: amfGrpcServer/amf_server.go [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+			}
+		}
+	}
+    return nil
+}
+
+func Start_service() {
+
+	fmt.Println("called Start_service ")
+	lis, err := net.Listen("tcp", ":9000")
```
Comment (thakurajayL) [inline amfGrpcServer/amf_server.go:65]: Make this configurable. Don't use fix value.

File: amfGrpcServer/amf_server.go [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+	var EndLoopChan chan bool
+
+	go func() {
+		for {
+			select {
+			case msg1 := <-Amf2RanMsgChan:
+			    log.Printf("Send Response message body from client (%s): Body - %s, MsgType %v ", msg1.ServerId, msg1.Body, msg1.Msgtype)
+				if err := srv.Send(msg1); err != nil {
+					log.Println("Error in sending response")
+				}
```
Comment (thakurajayL) [inline amfGrpcServer/amf_server.go:30]: If not used then just get rid of EndLoopChan

File: amfGrpcServer/amf_server.go [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+			} else {
+				ngap.DispatchLb("1.1.1.1", req.Msg, Amf2RanMsgChan)
+			}
+		}
+	}
+    return nil
+}
+
+func Start_service() {
```
Comment (thakurajayL) [inline amfGrpcServer/amf_server.go:62]: This call should be conditional based on if SCTP pod exist or not.

File: amfGrpcServer/amf_server.go (hunk 14 above)
Comment (thakurajayL) [reply amfGrpcServer/amf_server.go]: Resolved. Moved this file to service package.

File: amfGrpcServer/amf_server.go (hunk 16 above)
Comment (thakurajayL) [reply amfGrpcServer/amf_server.go]: Removed

File: context/context.go [hunk 18]
```diff
@@ -310,6 +313,22 @@ func (context *AMFContext) AmfRanFindByConn(conn net.Conn) (*AmfRan, bool) {
...
+func (context *AMFContext) NewAmfRanId(remoteAddr string) *AmfRan {
+	ran := AmfRan{}
+	ran.SupportedTAList = NewSupportedTAIList()
+	ran.GnbIp = remoteAddr
+	ran.Log = logger.NgapLog.WithField(logger.FieldRanAddr, remoteAddr)
+	context.AmfRanPool.Store(remoteAddr, &ran)
+	return &ran
+}
+
+func (context *AMFContext) AmfRanFindById(remoteAddr string) (*AmfRan, bool) {
```
Comment (thakurajayL) [inline context/context.go:325]: Change function name - FindByAddr

File: gmm/handler.go [hunk 19]
```diff
@@ -1294,7 +1294,7 @@ func handleRequestedNssai(ue *context.AmfUe, anType models.AccessType) error {
 					UserLocation:     &ue.Location,
 					RrcEstCause:      ue.RanUe[anType].RRCEstablishmentCause,
 					UeContextRequest: ue.RanUe[anType].UeContextRequest,
-					AnN2IPv4Addr:     ue.RanUe[anType].Ran.Conn.RemoteAddr().String(),
+					AnN2IPv4Addr:     ue.RanUe[anType].Ran.GnbIp,
```
Comment (thakurajayL) [inline gmm/handler.go:1297]: Is GnbIp and remoteaddr same ?

File: go.mod [hunk 20]
```diff
@@ -32,5 +32,7 @@ require (
 	github.com/sirupsen/logrus v1.8.1
 	github.com/spf13/viper v1.7.1
 	github.com/urfave/cli v1.22.4
+	google.golang.org/grpc v1.39.0
```
Comment (thakurajayL) [inline go.mod:35]: Why its showing as addition ? Isn't it something should already be there ?

File: ngap/dispatcher.go [hunk 21]
```diff
@@ -237,3 +286,29 @@ func HandleSCTPNotification(conn net.Conn, notification sctp.Notification) {
 		ran.Log.Warnf("Non handled notification type: 0x%x", notification.Type())
 	}
 }
+
+func HandleSCTPNotificationLb(remoteAddr string) {
```
Comment (thakurajayL) [inline ngap/dispatcher.go:290]: Formatting issues with this file ?

Comment (thakurajayL) [reply ngap/dispatcher.go:290]: Should see if can use interface.

File: service/amf_server.go [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func (s *Server) HandleMessage(srv sdcoreAmfServer.NgapService_HandleMessageServer) error {
+	var Amf2RanMsgChan chan *sdcoreAmfServer.Message
+	Amf2RanMsgChan = make(chan *sdcoreAmfServer.Message, 100)
+
+	go func() {
+		for {
+			select {
+			case msg1 := <-Amf2RanMsgChan:
```
Comment (thakurajayL) [inline service/amf_server.go:24]: Run with 2 instances of SCTPLB

File: context/context.go (hunk 18 above)
Comment (thakurajayL) [reply context/context.go:325]: Done

File: gmm/handler.go (hunk 19 above)
Comment (thakurajayL) [reply gmm/handler.go:1297]: RemoteAddress is not used anymore.

**Review by badhrinathpa — APPROVED** [review]
+1

**Review by badhrinathpa — APPROVED** [review]
+1

**Pull request merged**